model/task_lifecycle_test.go

```diff
@@ -3614,6 +3614,151 @@ tasks:
+		Status:              evergreen.TaskUndispatched,
+		Activated:           false,
+		RevisionOrderNumber: 2,
+	}
+	prevT2 := task.Task{
+		Id:                  "prev_t2",
+		BuildId:             "prev_b1",
+		Version:             "prev_v1",
+		TaskGroup:           "my_group",
+		DisplayName:         "second",
+		Requester:           evergreen.RepotrackerVersionRequester,
+		Status:              evergreen.TaskUndispatched,
+		Activated:           false,
+		RevisionOrderNumber: 2,
+	}
+	prevT3 := task.Task{
+		Id:                  "prev_t3",
+		BuildId:             "prev_b1",
+		Version:             "prev_v1",
+		TaskGroup:           "my_group",
+		DisplayName:         "third",
+		Requester:           evergreen.RepotrackerVersionRequester,
+		Status:              evergreen.TaskUndispatched,
+		Activated:           false,
+		RevisionOrderNumber: 2,
+	}
+	prevSuccessT1 := task.Task{
+		Id:                  "prev_success_t1",
+		BuildId:             "prev_success_b1",
+		Version:             "prev_success_v1",
+		TaskGroup:           "my_group",
+		DisplayName:         "first",
+		Requester:           evergreen.RepotrackerVersionRequester,
+		Status:              evergreen.TaskSucceeded,
+		Activated:           true,
+		RevisionOrderNumber: 1,
+	}
+	prevSuccessT2 := task.Task{
+		Id:                  "prev_success_t2",
+		BuildId:             "prev_success_b1",
+		Version:             "prev_success_v1",
+		TaskGroup:           "my_group",
+		DisplayName:         "second",
+		Requester:           evergreen.RepotrackerVersionRequester,
+		Status:              evergreen.TaskSucceeded,
+		Activated:           true,
+		RevisionOrderNumber: 1,
+	}
+	prevSuccessT3 := task.Task{
+		Id:                  "prev_success_t3",
+		BuildId:             "prev_success_b1",
+		Version:             "prev_success_v1",
+		TaskGroup:           "my_group",
+		DisplayName:         "third",
+		Requester:           evergreen.RepotrackerVersionRequester,
+		Status:              evergreen.TaskSucceeded,
+		Activated:           true,
+		RevisionOrderNumber: 1,
+	}
+	assert.NoError(t, db.InsertMany(task.Collection, t1, t2, t3, prevT1, prevT2, prevT3, prevSuccessT1, prevSuccessT2, prevSuccessT3))
+	assert.NoError(t, evalStepback(&t2, "", evergreen.TaskFailed, false))
+
+	// verify only the previous t1 and t2 are stepped back
+	prevT1FromDb, err := task.FindOneId(prevT1.Id)
+	assert.NoError(t, err)
+	assert.True(t, prevT1FromDb.Activated)
+	prevT2FromDb, err := task.FindOneId(prevT2.Id)
+	assert.NoError(t, err)
+	assert.True(t, prevT2FromDb.Activated)
+	prevT3FromDb, err := task.FindOneId(prevT3.Id)
+	assert.NoError(t, err)
+	assert.False(t, prevT3FromDb.Activated)
+
+	// stepping back t3 should now also stepback t3 and not error on earlier activated tasks
+	assert.NoError(t, evalStepback(&t3, "", evergreen.TaskFailed, false))
+	prevT3FromDb, err = task.FindOneId(prevT3.Id)
+	assert.NoError(t, err)
+	assert.True(t, prevT3FromDb.Activated)
+}
```

Comment:
It may be worthwhile to add some testing on "unhappy path" error assertions as well, i.e. an error returned when stepping back task group for an earlier task, or  len(tasks) == 0.

Reply:
There is a test for stepping back an earlier task in the task group (although this isn't an error case so I'm not sure if I'm misunderstanding that suggestion). The len(tasks) == 0 is really an impossible state that shouldn't happen.

Reply:
I was referring to the case on line 814 where doStepback for an earlier task can fail to ensure that's handled correctly.  Up to you if you think it's necessart.  LGTM.

Reply:
Gotcha-- doStepback has it's own tests so I don't think it needs to be handled specifically here.